File: DKPlugins/DKDuktape/DKEvent.js
// [MDN] https://developer.mozilla.org/en-US/docs/Web/API/Event
// [INTERFACE] https://dom.spec.whatwg.org/#interface-event

//event_instances = [];

// [Event()] https://developer.mozilla.org/en-US/docs/Web/API/Event/Event
var Event = function Event(_type, _options, _pointer) {
	
	/*
    this._type = _type;
	this._options = _options;
	this._pointer = _pointer;
	*/

    ////// Instance properties //////
	// [Event.bubbles](Read only) https://developer.mozilla.org/en-US/docs/Web/API/Event/bubbles
    Object.defineProperty(this, "bubbles", {
        get: function bubbles() {
            return CPP_DKEvent_bubbles(_pointer);
        }
    });
	// [Event.cancelable](Read only) https://developer.mozilla.org/en-US/docs/Web/API/Event/cancelable
    Object.defineProperty(this, "cancelable", {
        get: function cancelable() {
            return CPP_DKEvent_cancelable(_pointer);
        }
    });
    // [Event.composed](Read only) https://developer.mozilla.org/en-US/docs/Web/API/Event/composed
    Object.defineProperty(this, "composed", {
        get: function composed() {
            return CPP_DKEvent_composed(_pointer);
        }
    });
    // [Event.currentTarget](Read only) https://developer.mozilla.org/en-US/docs/Web/API/Event/currentTarget
    Object.defineProperty(this, "currentTarget", {
        get: function currentTarget() {
			return window;
			/*
            var elementPointer = CPP_DKEvent_currentTarget(_pointer);
            if (!elementPointer) {
                return;
            }
            return new HTMLElement(elementPointer);
			*/
        }
    });
	// [Event.defaultPrevented](Read only) https://developer.mozilla.org/en-US/docs/Web/API/Event/defaultPrevented
	Object.defineProperty(this, "defaultPrevented", {
        get: function defaultPrevented() {
            return CPP_DKEvent_defaultPrevented(_pointer);
        }
    });
	// [Event.eventPhase](Read only) https://developer.mozilla.org/en-US/docs/Web/API/Event/eventPhase
	Object.defineProperty(this, "eventPhase", {
        get: function eventPhase() {
            return CPP_DKEvent_eventPhase(_pointer);
        }
    });
	// [Event.isTrusted](Read only) https://developer.mozilla.org/en-US/docs/Web/API/Event/isTrusted
	Object.defineProperty(this, "isTrusted", {
        get: function isTrusted() {
            return CPP_DKEvent_isTrusted(_pointer);
        }
    });
	// [Event.target](Read only) https://developer.mozilla.org/en-US/docs/Web/API/Event/target
	/*
    Object.defineProperty(this, "target", {
        //Read Only
        get: function target() {
            var elementPointer = CPP_DKEvent_target(_pointer);
            if (!elementPointer) {
                return;
            }
            return new HTMLElement(elementPointer);
        }
    });
	*/
	// [Event.timeStamp](Read only) https://developer.mozilla.org/en-US/docs/Web/API/Event/timeStamp
	Object.defineProperty(this, "timeStamp", {
        get: function timeStamp() {
            return CPP_DKEvent_timeStamp(_pointer);
        }
    });
	// [Event.type](Read only) https://developer.mozilla.org/en-US/docs/Web/API/Event/type
	Object.defineProperty(this, "type", {
        get: function type() {
            return this._type//CPP_DKEvent_type(_pointer);
        }
    });
    
    ////// Legacy and non-standard properties //////
	// [Event.cancelBubble](Read only) https://developer.mozilla.org/en-US/docs/Web/API/Event/cancelBubble
    Object.defineProperty(this, "cancelBubble", {
        set: function cancelBubble(flag) {
            return CPP_DKEvent_cancelBubble(_pointer, flag);
        }
    });
	// [Event.explicitOriginalTarget](Read only)(Non-standard) https://developer.mozilla.org/en-US/docs/Web/API/Event/explicitOriginalTarget
	Object.defineProperty(this, "explicitOriginalTarget", {
        get: function explicitOriginalTarget() {
            return CPP_DKEvent_explicitOriginalTarget(_pointer);
        }
    });
	// [Event.originalTarget](Read only)(Non-standard) https://developer.mozilla.org/en-US/docs/Web/API/Event/originalTarget
	Object.defineProperty(this, "originalTarget", {
        get: function originalTarget() {
            return CPP_DKEvent_originalTarget(_pointer);
        }
    });
	// [Event.originalTarget](Deprecated) https://developer.mozilla.org/en-US/docs/Web/API/Event/returnValue
	Object.defineProperty(this, "returnValue", {
        get: function returnValue() {
            return CPP_DKEvent_returnValue(_pointer);
        }
    });
	// [Event.scoped](Read only)(Deprecated)
	Object.defineProperty(this, "scoped", {
        get: function scoped() {
            return CPP_DKEvent_scoped(_pointer);
        }
    });
	
	////// Obsolete properties //////
	/*
	Object.defineProperty(this, "deepPath", {
        get: function deepPath(){
            return CPP_DKEvent_deepPath(_pointer);
        }
    });
    Object.defineProperty(this, "srcElement", {
        get: function srcElement() {
            var elementPointer = CPP_DKEvent_srcElement(_pointer);
            if (!elementPointer) {
                return;
            }
            return new HTMLElement(elementPointer);
        }
    });
	*/

    ////// Instance methods //////
	// [Event.composedPath()] https://developer.mozilla.org/en-US/docs/Web/API/Event/composedPath
	Event.prototype.composedPath = function composedPath() {
        //TODO
    }
	// [Event.preventDefault()] https://developer.mozilla.org/en-US/docs/Web/API/Event/preventDefault
	Event.prototype.preventDefault = function preventDefault() {
        CPP_DKEvent_preventDefault(_pointer);
    }
	// [Event.stopImmediatePropagation()] https://developer.mozilla.org/en-US/docs/Web/API/Event/stopImmediatePropagation
	Event.prototype.stopImmediatePropagation = function stopImmediatePropagation() {
        CPP_DKEvent_stopImmediatePropagation(_pointer);
    }
	// [Event.stopPropagation()] https://developer.mozilla.org/en-US/docs/Web/API/Event/stopPropagation
	Event.prototype.stopPropagation = function stopPropagation() {
        CPP_DKEvent_stopPropagation(_pointer);
    }
	
	////// Deprecated methods //////
	// [Event.initEvent()](Deprecated) https://developer.mozilla.org/en-US/docs/Web/API/Event/initEvent
	Event.prototype.initEvent = function initEvent() {
        CPP_DKEvent_initEvent(_pointer);
    }
	
	////// Obsolete methods //////
	/*
    Event.prototype.createEvent = function createEvent() {
        CPP_DKEvent_createEvent(_pointer);
    }
    Event.prototype.composedPath = function composedPath() {
        CPP_DKEvent_composedPath(_pointer);
    }
    Event.prototype.getPreventDefault = function getPreventDefault() {
        CPP_DKEvent_getPreventDefault(_pointer);
    }
    Event.prototype.preventBubble = function preventBubble() {
        CPP_DKEvent_preventBubble(_pointer);
    }
    Event.prototype.preventCapture = function preventCapture() {
        CPP_DKEvent_preventCapture(_pointer);
    }
	*/

	////// toString //////
	if(this.toString() === "[object Object]"){
		this.toString = function(){
			return "[object Event]"
		}
	}
	
    //console.log("dispatching event");
    //this.currentTarget.dispatchEvent(this);
    //return this;
}

// Called from C++ to send events
///////////////////////////////
function DispatchEvent(_pointer) {
    var event = new Event(_pointer);
    //console.log("DispatchEvent("+pointer+"): event.type = "+event.type);

    if (event.type === "mousemove" || event.type === "mouseover" || event.type === "mousedown" || event.type === "mouseup" || event.type === "click" || event.type === "dblclick" || event.type === "contextmenu") {
        var mouseEvent = new MouseEvent(_type, _options, _pointer);
        mouseEvent.currentTarget.dispatchEvent(mouseEvent);
    } else if (event.type === "keydown" || event.type === "keyup" || event.type === "keypress") {
        var keyboardEvent = new KeyboardEvent(pointer);
        keyboardEvent.currentTarget.dispatchEvent(keyboardEvent);
    } else {
        event.currentTarget.dispatchEvent(event);
    }
}

/////////////////////////////////////////////
function doEvent(_type, _options, _pointer) {
	const event = new Event(_type, _options, _pointer)
	event.target = new EventTarget(_pointer)
	event.target.dispatchEvent(event)
}

File: DKPlugins/DKDuktape/DKEvent.cpp
// https://developer.mozilla.org/en-US/docs/Web/API/Event

#include "DK/DKApp.h"
#include "DKDuktape/DKEvent.h"
//#include "DKDom/DKElement.h"


///////////////////////
bool DKEvent::Init()
{
	DKDEBUGFUNC();

	// Properties
	DKDuktape::AttachFunction("CPP_DKEvent_bubbles", DKEvent::bubbles);
	DKDuktape::AttachFunction("CPP_DKEvent_cancelBubble", DKEvent::cancelBubble);
	DKDuktape::AttachFunction("CPP_DKEvent_cancelable", DKEvent::cancelable);
	DKDuktape::AttachFunction("CPP_DKEvent_composed", DKEvent::composed);
	DKDuktape::AttachFunction("CPP_DKEvent_currentTarget", DKEvent::currentTarget);
	DKDuktape::AttachFunction("CPP_DKEvent_deepPath", DKEvent::deepPath);
	DKDuktape::AttachFunction("CPP_DKEvent_defaultPrevented", DKEvent::defaultPrevented);
	DKDuktape::AttachFunction("CPP_DKEvent_eventPhase", DKEvent::eventPhase);
	DKDuktape::AttachFunction("CPP_DKEvent_explicitOriginalTarget", DKEvent::explicitOriginalTarget);
	DKDuktape::AttachFunction("CPP_DKEvent_originalTarget", DKEvent::originalTarget);
	DKDuktape::AttachFunction("CPP_DKEvent_returnValue", DKEvent::returnValue);
	DKDuktape::AttachFunction("CPP_DKEvent_srcElement", DKEvent::srcElement);
	DKDuktape::AttachFunction("CPP_DKEvent_target", DKEvent::target);
	DKDuktape::AttachFunction("CPP_DKEvent_timeStamp", DKEvent::timeStamp);
	DKDuktape::AttachFunction("CPP_DKEvent_type", DKEvent::type);
	DKDuktape::AttachFunction("CPP_DKEvent_isTrusted", DKEvent::isTrusted);

	// Obsolete properties
	DKDuktape::AttachFunction("CPP_DKEvent_scoped", DKEvent::scoped);

	// Methods
	DKDuktape::AttachFunction("CPP_DKEvent_createEvent", DKEvent::createEvent);
	DKDuktape::AttachFunction("CPP_DKEvent_composedPath", DKEvent::composedPath);
	DKDuktape::AttachFunction("CPP_DKEvent_initEvent", DKEvent::initEvent);
	DKDuktape::AttachFunction("CPP_DKEvent_preventDefault", DKEvent::preventDefault);
	DKDuktape::AttachFunction("CPP_DKEvent_stopImmediatePropagation", DKEvent::stopImmediatePropagation);
	DKDuktape::AttachFunction("CPP_DKEvent_stopPropagation", DKEvent::stopPropagation);

	// Obsolete methods
	DKDuktape::AttachFunction("CPP_DKEvent_getPreventDefault", DKEvent::getPreventDefault);
	DKDuktape::AttachFunction("CPP_DKEvent_preventBubble", DKEvent::preventBubble);
	DKDuktape::AttachFunction("CPP_DKEvent_preventCapture", DKEvent::preventCapture);

	// Extra
	DKDuktape::AttachFunction("CPP_DKEvent_getParameters", DKEvent::getParameters);

	DKClass::DKCreate("DKDuktape/DKEvent.js");
	return true;
}


// Properties
//////////////////////////////////////
int DKEvent::bubbles(duk_context* ctx)
{
	DKDEBUGFUNC(ctx);
	/*
	DKString eventAddress = duk_require_string(ctx, 0);
	Rml::Event* event = DKRml::addressToEvent(eventAddress);
	if(!event){
		DKERROR("DKEvent::bubbles(): event invalid\n");
		duk_push_boolean(ctx, false);
		return true;
	}
	int phase = (int)event->GetPhase();
	if(phase == 4){
		duk_push_boolean(ctx, true);
		return true;
	}
	duk_push_boolean(ctx, false);
	*/
	return true;
}

//////////////////////////////////////////////
int DKEvent::cancelBubble(duk_context* ctx)
{
	DKDEBUGFUNC(ctx);
	/*
	DKString eventAddress = duk_require_string(ctx, 0);
	Rml::Event* event = DKRml::addressToEvent(eventAddress);
	if (!event) {
		DKERROR("DKEvent::cencelBubble(): event invalid\n");
		duk_push_boolean(ctx, false);
		return true;
	}
	bool flag = duk_require_boolean(ctx, 1);
	if(flag){
		event->StopPropagation();
		return true;
	}
	*/
	return true;
}

////////////////////////////////////////////
int DKEvent::cancelable(duk_context* ctx)
{
	DKDEBUGFUNC(ctx);
	/*
	DKString eventAddress = duk_require_string(ctx, 0);
	Rml::Event* event = DKRml::addressToEvent(eventAddress);
	if (!event) {
		DKERROR("DKEvent::cancelable(): event invalid\n");
		duk_push_boolean(ctx, false);
		return true;
	}

	//TODO
	*/
	return false;
}

//////////////////////////////////////////
int DKEvent::composed(duk_context* ctx)
{
	DKDEBUGFUNC(ctx);
	/*
	DKString eventAddress = duk_require_string(ctx, 0);
	Rml::Event* event = DKRml::addressToEvent(eventAddress);
	if (!event) {
		DKERROR("DKEvent::composed(): event invalid\n");
		duk_push_boolean(ctx, false);
		return true;
	}

	//TODO
	*/
	return false;
}

///////////////////////////////////////////////
int DKEvent::currentTarget(duk_context* ctx)
{
	DKDEBUGFUNC(ctx);
	/*
	DKString eventAddress = duk_require_string(ctx, 0);
	Rml::Event* event = DKRml::addressToEvent(eventAddress);
	if (!event) {
		DKERROR("DKEvent::currentTarget(): event invalid\n");
		duk_push_boolean(ctx, false);
		return true;
	}

	Rml::Element* currentElement = event->GetCurrentElement();
	DKString currentElementAddress = DKRml::elementToAddress(currentElement);
	if(currentElementAddress.empty()){ 
		DKERROR("DKEvent::currentTarget(): currentElementAddress invalid\n");
		return true; 
	}
	duk_push_string(ctx, currentElementAddress.c_str());
	*/
	return true;
}

//////////////////////////////////////////
int DKEvent::deepPath(duk_context* ctx)
{
	DKDEBUGFUNC(ctx);
	/*
	DKString eventAddress = duk_require_string(ctx, 0);
	Rml::Event* event = DKRml::addressToEvent(eventAddress);
	if (!event) {
		DKERROR("DKEvent::deepPath(): event invalid\n");
		duk_push_boolean(ctx, false);
		return true;
	}

	//TODO
	*/
	return false;
}

//////////////////////////////////////////////////
int DKEvent::defaultPrevented(duk_context* ctx)
{
	DKDEBUGFUNC(ctx);
	/*
	DKString eventAddress = duk_require_string(ctx, 0);
	Rml::Event* event = DKRml::addressToEvent(eventAddress);
	if (!event) {
		DKERROR("DKEvent::defaultPrevented(): event invalid\n");
		duk_push_boolean(ctx, false);
		return true;
	}

	//TODO
	*/
	return false;
}

////////////////////////////////////////////
int DKEvent::eventPhase(duk_context* ctx)
{
	DKDEBUGFUNC(ctx);
	/*
	DKString eventAddress = duk_require_string(ctx, 0);
	Rml::Event* event = DKRml::addressToEvent(eventAddress);
	if (!event) {
		DKERROR("DKEvent::eventPhase(): event invalid\n");
		duk_push_boolean(ctx, false);
		return true;
	}
	int phase = (int)event->GetPhase();
	duk_push_int(ctx, phase);
	*/
	return true;
}

////////////////////////////////////////////////////////
int DKEvent::explicitOriginalTarget(duk_context* ctx)
{
	DKDEBUGFUNC(ctx);
	/*
	DKString eventAddress = duk_require_string(ctx, 0);
	Rml::Event* event = DKRml::addressToEvent(eventAddress);
	if (!event) {
		DKERROR("DKEvent::explicitOriginalTarget(): event invalid\n");
		duk_push_boolean(ctx, false);
		return true;
	}

	//TODO
	*/
	return false;
}

////////////////////////////////////////////////
int DKEvent::originalTarget(duk_context* ctx)
{
	DKDEBUGFUNC(ctx);
	/*
	DKString eventAddress = duk_require_string(ctx, 0);
	Rml::Event* event = DKRml::addressToEvent(eventAddress);
	if (!event) {
		DKERROR("DKEvent::originalTarget(): event invalid\n");
		duk_push_boolean(ctx, false);
		return true;
	}

	//TODO
	*/
	return false;
}

/////////////////////////////////////////////
int DKEvent::returnValue(duk_context* ctx)
{
	DKDEBUGFUNC(ctx);
	/*
	DKString eventAddress = duk_require_string(ctx, 0);
	Rml::Event* event = DKRml::addressToEvent(eventAddress);
	if (!event) {
		DKERROR("DKEvent::returnValue(): event invalid\n");
		duk_push_boolean(ctx, false);
		return true;
	}

	//TODO
	*/
	return false;
}

////////////////////////////////////////////
int DKEvent::srcElement(duk_context* ctx)
{
	DKDEBUGFUNC(ctx);
	/*
	DKString eventAddress = duk_require_string(ctx, 0);
	Rml::Event* event = DKRml::addressToEvent(eventAddress);
	if (!event) {
		DKERROR("DKEvent::srcElement(): event invalid\n");
		duk_push_boolean(ctx, false);
		return true;
	}
	Rml::Element* srcElement = event->GetCurrentElement();
	DKString srcElementAddress = DKRml::elementToAddress(srcElement);
	if (srcElementAddress.empty()) {
		DKERROR("DKEvent::srcElement(): srcElementAddress invalid\n");
		return true;
	}
	duk_push_string(ctx, srcElementAddress.c_str());
	*/
	return true;
}

////////////////////////////////////////
int DKEvent::target(duk_context* ctx)
{
	DKDEBUGFUNC(ctx);
	/*
	DKString eventAddress = duk_require_string(ctx, 0);
	Rml::Event* event = DKRml::addressToEvent(eventAddress);
	if (!event) {
		DKERROR("DKEvent::target(): event invalid\n");
		duk_push_boolean(ctx, false);
		return true;
	}
	Rml::Element* targetElement = event->GetCurrentElement();
	DKString targetElementAddress = DKRml::elementToAddress(targetElement);
	if (targetElementAddress.empty()) {
		DKERROR("DKEvent::target(): targetElementAddress invalid\n");
		return true;
	}
	duk_push_string(ctx, targetElementAddress.c_str());
	*/
	return true;
}

///////////////////////////////////////////
int DKEvent::timeStamp(duk_context* ctx)
{
	DKDEBUGFUNC(ctx);
	/*
	DKString eventAddress = duk_require_string(ctx, 0);
	Rml::Event* event = DKRml::addressToEvent(eventAddress);
	if (!event) {
		DKERROR("DKEvent::timeStamp(): event invalid\n");
		duk_push_boolean(ctx, false);
		return true;
	}

	//TODO*/
	return false;
}

//////////////////////////////////////
int DKEvent::type(duk_context* ctx)
{
	DKDEBUGFUNC(ctx);
	/*
	DKString eventAddress = duk_require_string(ctx, 0);
	Rml::Event* event = DKRml::addressToEvent(eventAddress);
	if (!event) {
		DKERROR("DKEvent::type(): event invalid\n");
		duk_push_boolean(ctx, false);
		return true;
	}
	DKString type = event->GetType();
	if(type.empty()){
		DKERROR("DKEvent::type(): event type invalid\n");
		return true; 
	}

	if(same(type, "mouseup") && event->GetParameter<int>("button", 0) == 1){
		type = "contextmenu";
	}

	duk_push_string(ctx, type.c_str());
	*/
	return true;
}

///////////////////////////////////////////
int DKEvent::isTrusted(duk_context* ctx)
{
	DKDEBUGFUNC(ctx);
	/*
	DKString eventAddress = duk_require_string(ctx, 0);
	Rml::Event* event = DKRml::addressToEvent(eventAddress);
	if (!event) {
		DKERROR("DKEvent::isTrusted(): event invalid\n");
		duk_push_boolean(ctx, false);
		return true;
	}

	//TODO
	*/
	return false;
}


// Obsolete properties
////////////////////////////////////////
int DKEvent::scoped(duk_context* ctx)
{
	DKDEBUGFUNC(ctx);
	/*
	DKString eventAddress = duk_require_string(ctx, 0);
	Rml::Event* event = DKRml::addressToEvent(eventAddress);
	if (!event) {
		DKERROR("DKEvent::scoped(): event invalid\n");
		duk_push_boolean(ctx, false);
		return true;
	}

	//TODO
	*/
	return false;
}


// Methods
/////////////////////////////////////////////
int DKEvent::createEvent(duk_context* ctx) 
{
	DKDEBUGFUNC(ctx);
	/*
	DKString eventAddress = duk_require_string(ctx, 0);
	Rml::Event* event = DKRml::addressToEvent(eventAddress);
	if (!event) {
		DKERROR("DKEvent::createEvent(): event invalid\n");
		duk_push_boolean(ctx, false);
		return true;
	}

	//TODO
	*/
	return false;
}

//////////////////////////////////////////////
int DKEvent::composedPath(duk_context* ctx)
{
	DKDEBUGFUNC(ctx);
	/*
	DKString eventAddress = duk_require_string(ctx, 0);
	Rml::Event* event = DKRml::addressToEvent(eventAddress);
	if (!event) {
		DKERROR("DKEvent::composedPath(): event invalid\n");
		duk_push_boolean(ctx, false);
		return true;
	}

	//TODO
	*/
	return false;
}

///////////////////////////////////////////
int DKEvent::initEvent(duk_context* ctx)
{
	DKDEBUGFUNC(ctx);
	/*
	DKString eventAddress = duk_require_string(ctx, 0);
	Rml::Event* event = DKRml::addressToEvent(eventAddress);
	if (!event) {
		DKERROR("DKEvent::initEvent(): event invalid\n");
		duk_push_boolean(ctx, false);
		return true;
	}

	//TODO
	*/
	return false;
}

////////////////////////////////////////////////
int DKEvent::preventDefault(duk_context* ctx)
{
	DKDEBUGFUNC(ctx);
	/*
	DKString eventAddress = duk_require_string(ctx, 0);
	Rml::Event* event = DKRml::addressToEvent(eventAddress);
	if (!event) {
		DKERROR("DKEvent::preventDefault(): event invalid\n");
		duk_push_boolean(ctx, false);
		return true;
	}
	//TODO
	*/
	return false;
}

//////////////////////////////////////////////////////////
int DKEvent::stopImmediatePropagation(duk_context* ctx)
{
	DKDEBUGFUNC(ctx);
	/*
	DKString eventAddress = duk_require_string(ctx, 0);
	Rml::Event* event = DKRml::addressToEvent(eventAddress);
	if(!event){
		DKERROR("DKEvent::stopImmediatePropagation(): event invalid\n");
		duk_push_boolean(ctx, false);
		return true;
	}
	event->StopImmediatePropagation();
	*/
	return true;
}

/////////////////////////////////////////////////
int DKEvent::stopPropagation(duk_context* ctx)
{
	DKDEBUGFUNC(ctx);
	/*
	DKString eventAddress = duk_require_string(ctx, 0);
	Rml::Event* event = DKRml::addressToEvent(eventAddress);
	if(!event){
		DKERROR("DKEvent::stopPropagation(): event invalid\n");
		duk_push_boolean(ctx, false);
		return true;
	}
	event->StopPropagation();
	*/
	return true;
}


// Obsolete methods
///////////////////////////////////////////////////
int DKEvent::getPreventDefault(duk_context* ctx)
{
	DKDEBUGFUNC(ctx);
	/*
	DKString eventAddress = duk_require_string(ctx, 0);
	Rml::Event* event = DKRml::addressToEvent(eventAddress);
	if (!event) {
		DKERROR("DKEvent::getPreventDefault(): event invalid\n");
		duk_push_boolean(ctx, false);
		return true;
	}

	//TODO
	*/
	return false;
}

///////////////////////////////////////////////
int DKEvent::preventBubble(duk_context* ctx)
{
	DKDEBUGFUNC(ctx);
	/*
	DKString eventAddress = duk_require_string(ctx, 0);
	Rml::Event* event = DKRml::addressToEvent(eventAddress);
	if (!event) {
		DKERROR("DKEvent::preventBubble(): event invalid\n");
		duk_push_boolean(ctx, false);
		return true;
	}

	//TODO
	*/
	return false;
}

////////////////////////////////////////////////
int DKEvent::preventCapture(duk_context* ctx)
{
	DKDEBUGFUNC(ctx);
	/*
	DKString eventAddress = duk_require_string(ctx, 0);
	Rml::Event* event = DKRml::addressToEvent(eventAddress);
	if (!event) {
		DKERROR("DKEvent::preventCapture(): event invalid\n");
		duk_push_boolean(ctx, false);
		return true;
	}

	//TODO
	*/
	return false;
}


// Extra
///////////////////////////////////////////////
int DKEvent::getParameters(duk_context* ctx)
{
	DKDEBUGFUNC(ctx);
	/*
	DKString eventAddress = duk_require_string(ctx, 0);
	Rml::Event* event = DKRml::addressToEvent(eventAddress);
	if (!event) {
		DKERROR("DKEvent::getParameters(): event invalid\n");
		duk_push_boolean(ctx, false);
		return true;
	}

	//List event parameters and values
	DKINFO("LISTING EVENT PARAMETERS\n");
	const auto& p = event->GetParameters();
	DKString output = "";
	for(auto& entry : p){
		output += entry.first + ": " + entry.second.Get<Rml::String>() + "\n";
	}
	DKINFO(output);
	duk_push_string(ctx, output.c_str());
	*/
	return true;
}